server/src/models/timestamps.rs

```diff
@@ -0,0 +1,352 @@
+//! been collected at a regular sampling interval the residual timestamps are
+//! compressed as the segment's length with the prefix zero bits stripped,
+//! otherwise the compression method proposed for timestamps for the time series
+//! management system Gorilla in the [Gorilla paper] is used.
+//!
+//! [Gorilla paper]: https://www.vldb.org/pvldb/vol8/p1816-teller.pdf
+
+use std::mem;
+
+use crate::compression;
+use crate::models::bits::{BitReader, BitVecBuilder};
+use crate::types::{Timestamp, TimestampArray, TimestampBuilder};
+
+/// Compress the timestamps in `uncompressed_timestamps` from the second
+/// timestamp to the second to last timestamp. The first and last timestamp are
+/// already stored as part of the compressed segment to allow segments to be
+/// pruned. If the time series is regular, the timestamps are encoded as the
+/// number of data points in the segments with prefix zeros stripped, and if it
+/// is irregular, the timestamp's delta-of-delta is computed and then encoded
+/// using a variable length binary encoding.
+pub fn compress_residual_timestamps(uncompressed_timestamps: &TimestampArray) -> Vec<u8> {
+    // Nothing to do as the segments already store the first and last timestamp.
+    if uncompressed_timestamps.len() <= 2 {
+        return vec![];
+    }
+
+    // Compress the residual timestamps using an optimized method depending on
+    // if the segment the timestamps are from is regular or irregular.
+    if are_timestamps_regular(uncompressed_timestamps.values()) {
+        // The timestamps are regular, so only the segment's length is stored as
+        // an integer with all the prefix zeros stripped from the integer.
+        compress_regular_residual_timestamps(uncompressed_timestamps)
+    } else {
+        // The timestamps are irregular so they are compressed as
+        // delta-of-deltas stored using a variable length binary encoding.
+        compress_irregular_residual_timestamps(uncompressed_timestamps)
+    }
+}
+
+/// Return `true` if the timestamps in `uncompressed_timestamps` follow a
+/// regular sampling interval, otherwise `false`.
+fn are_timestamps_regular(uncompressed_timestamps: &[Timestamp]) -> bool {
+    if uncompressed_timestamps.len() < 2 {
+        return true;
+    }
+
+    // Unwrap is safe as uncompressed_timestamps contains at least two timestamps.
+    let expected_sampling_interval = uncompressed_timestamps[1] - uncompressed_timestamps[0];
+    let mut uncompressed_timestamps = uncompressed_timestamps.iter();
+    let mut previous_timestamp = uncompressed_timestamps.next().unwrap();
+
+    for current_timestamp in uncompressed_timestamps {
+        if current_timestamp - previous_timestamp != expected_sampling_interval {
+            return false;
+        }
+        previous_timestamp = current_timestamp;
+    }
+    true
+}
+
+/// Compress `uncompressed_timestamps` for a regular time series as a zero bit
```

Comment:
Maybe mention in the doc comment for `compress_residual_timestamps` that we have a flag to determine which compression method is used.

Reply:
Fixed.